docs/proposals/20231122-security-groups.md

```diff
@@ -0,0 +1,225 @@
+
+#### As an operator, I want to be able to manage additional security groups which can be used in the Machine Template
+
+As a cluster operator I want to provide additional security groups that can be used in the Machine Template and not to worry about the lifecycle of these security groups and the resources that are using them.
+
+#### As an operator, I want to be able to manage additional security group rules for the Bastion, the Control Plane and the Worker Nodes
+
+As a cluster operator I want to be able to specify more rules than the defaults, to have more flexibility in my cluster.
+
+#### As an operator, I want to be able to manage additional security group rules for the CNI
+
+As a cluster operator I want to have more options than Calico and provide my CNI's specific security group rules to have more flexibility in my cluster.
+
+### Implementation Details/Notes/Constraints
+
+The plan is to create a new controller and migrate the management of security groups and their rules in there when the `managedSecurityGroups` is set to `True`.
+
+A cluster operator will be able to add additional security groups and rules to the cluster by adding them to the `OpenStackCluster` spec.
+The controller will then reconcile the security groups and their rules to the cluster between what's in the spec and what's in the status.
+Special care will be taken to not remove security groups that are still in use by the machines.
+Also, during an upgrade, the controller will make sure to add the new security groups and their rules to the machines before reducing the default rules
+which won't have the Calico rules anymore.
+
+#### Data model changes
+
+The type for `OpenStackCluster.Spec.ManagedSecurityGroups` was a boolean but will become a structure with new fields.
+
+Example `ManagedSecurityGroups` yaml:
+
+```yaml
+managedSecurityGroups:
+  # Enable the management of security groups with the default rules (kubelet, etcd, etc.)
+  # The default stays false
+  enabled: true
+  # Allow to provide additional security group rules for the Bastion,
+  # the Control Plane and the Worker security groups
+  additionalBastionSecurityGroupRules:
+    - direction: ingress
+      ethertype: IPv4
+      portRangeMin: 1234
+      portRangeMax: 1234
+      protocol: tcp
+  additionalControlPlaneSecurityGroupRules:
+    - direction: ingress
+      ethertype: IPv4
+      portRangeMin: 1234
+      portRangeMax: 1234
+      protocol: tcp
+  additionalWorkerSecurityGroupRules:
+    - direction: ingress
+      ethertype: IPv4
+      portRangeMin: 1234
+      portRangeMax:1234
+      protocol: tcp
+  # Allow to provide rules for the CNI that will be applied to all nodes
+  cniSecurityGroupRules:
+    - direction: ingress
+      ethertype: IPv4
+      portRangeMin: 1234
+      portRangeMax: 1234
+      protocol: tcp
```

Comment:
Why do we call this one CNI, when it's basically a thing gathering rules added to each of the nodes? I'd abstain from designing this against a CNI. From the technical standpoint I think it's possible to use different CNIs on different nodes. It's totally not supported at the moment, but might be one day.

Reply:
I can name it `allNodesSecurityGroupRules`, I don't mind.